File: easy/1d-spreadsheet.py
import sys
import math

def parse_arg(arg):
    if arg == "_": return (None, None)
    if arg.startswith("$"): return ("ref", int(arg[1:]))
    return ("num", int(arg))

n = int(input())
cells = []
for i in range(n):
    operation, arg_1, arg_2 = input().split()
    c = {"op": operation, "a": parse_arg(arg_1), "b": parse_arg(arg_2)}
    cells.append(c)

def get_arg_value(arg) -> int:
    (t, e) = arg
    if t == "num": return e
    if t == "ref": return get_cell_value(cells[e])
    raise Exception()

def get_cell_value(c):
    if "val" in c: return c["val"]
    if c["op"] == "VALUE": c["val"] = get_arg_value(c["a"])                        ; return c["val"]
    if c["op"] == "ADD":   c["val"] = get_arg_value(c["a"]) + get_arg_value(c["b"]); return c["val"]
    if c["op"] == "SUB":   c["val"] = get_arg_value(c["a"]) - get_arg_value(c["b"]); return c["val"]
    if c["op"] == "MULT":  c["val"] = get_arg_value(c["a"]) * get_arg_value(c["b"]); return c["val"]
    raise Exception()

for i in range(n):
    print(get_cell_value(cells[i]))

File: easy/1d-spreadsheet.rs
use std::{io};

macro_rules! parse_input {
    ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

enum Value {
    Number(i32),
    Reference(usize),
    Nil,
}

enum Operation {
    Value, Add, Sub, Mult,
}

struct Cell{
    operation: Operation,
    a: Value,
    b: Value,
    value: std::cell::Cell<Option<i32>>,
}

fn parse_arg(a: &String) -> Value {
    if a == "_" { Value::Nil}
    else if a.contains("$") { Value::Reference(a[1..].parse::<usize>().unwrap())}
    else {Value::Number(a.parse::<i32>().unwrap())}
}

impl Cell {
    fn new(operation: &String, a: &String, b: &String) -> Cell
    {
        let pa = parse_arg(&a);
        let pb = parse_arg(&b);
        let op = match operation.as_str() {
            "VALUE" => { Operation::Value },
            "ADD" =>   { Operation::Add   },
            "SUB" =>   { Operation::Sub   },
            "MULT" =>  { Operation::Mult  },
            _ => panic!()
        };
        Cell{ operation: op, a: pa, b: pb, value: std::cell::Cell::new(Option::None)}
    }

    fn get_value(&self, cells: &Vec<Cell>) -> i32 {
        let val = match self.operation {
            Operation::Value => { self.a.get_value(cells)                                     },
            Operation::Add =>   { self.a.get_value(cells) + self.b.get_value(cells)},
            Operation::Sub =>   { self.a.get_value(cells) - self.b.get_value(cells)},
            Operation::Mult =>  { self.a.get_value(cells) * self.b.get_value(cells)},
        };
        self.value.set(Some(val));
        val
    }
}

impl Value {
    fn get_value(&self, cells: &Vec<Cell>) -> i32
    {
        match *self {
            Value::Number(num) => num,
            Value::Reference(i) => {cells[i].get_value(cells)},
            Value::Nil => panic!(),
        }
    }
}

fn main() {
    let mut input_line = String::new();
    io::stdin().read_line(&mut input_line).unwrap();
    let n = parse_input!(input_line, i32);
    let mut cells = Vec::<Cell>::new();
    for _ in 0..n as usize {
        let mut input_line = String::new();
        io::stdin().read_line(&mut input_line).unwrap();
        let inputs = input_line.split(" ").collect::<Vec<_>>();
        let operation = inputs[0].trim().to_string();
        let arg_1 = inputs[1].trim().to_string();
        let arg_2 = inputs[2].trim().to_string();

        let cell = Cell::new(&operation, &arg_1, &arg_2);
        cells.push(cell);
    }

    //let (mut hit, mut mis) = (0, 0);
    for c in cells.iter() {
        println!("{:?}", match c.value.get() {
            Some(val) => {/*hit+=1;*/ val},
            None => {/*mis+=1;*/ c.get_value(&cells)},
        });
    }
    //eprintln!("hit: {}, mis: {}", hit, mis)
}

